File: core/src/main/java/com/lomekwi/cave/timeline/SegmentGroup.java
package com.lomekwi.cave.timeline;

import com.lomekwi.cave.app.copy.Copyable;
import com.lomekwi.cave.app.selection.Selectable;

import java.io.Serial;
import java.io.Serializable;
import java.util.AbstractCollection;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * 一组被同时操作（选择/拖动/分割）的片段，实现 {@link Collection}{@code <Segment>}，
 */
public class SegmentGroup extends AbstractCollection<Segment> implements Serializable, Selectable, Copyable {
    @Serial
    private static final long serialVersionUID = 1L;
    private final Set<Segment> segments = new LinkedHashSet<>();
    private transient boolean selected;

    @Override
    public boolean add(Segment segment) {
        if (segments.add(segment)) {
            segment.setGroup(this);
            return true;
        }
        return false;
    }

    @Override
    public boolean remove(Object o) {
        if (o instanceof Segment seg && segments.remove(seg)) {
            seg.setGroup(null);
            return true;
        }
        return false;
    }

    @Override
    public void clear() {
        for (Segment s : segments) s.setGroup(null);
        segments.clear();
    }

    @Override
    public boolean contains(Object o) {
        return segments.contains(o);
    }

    @Override
    public boolean isEmpty() {
        return segments.isEmpty();
    }

    @Override
    public Iterator<Segment> iterator() {
        return segments.iterator();
    }

    @Override
    public int size() {
        return segments.size();
    }

    public Set<Segment> getSegments() {
        return Collections.unmodifiableSet(segments);
    }

    @Override
    public boolean isSelected() {
        return selected;
    }

    @Override
    public void setSelected(boolean selected) {
        this.selected = selected;
        for (Segment seg : segments) {
            seg.setSelected(selected);
        }
    }

    @Override
    public Copyable copy() {
        SegmentGroup newGroup = new SegmentGroup();
        for (Segment seg : segments) {
            var dup = seg.duplicate();
            dup.setTrack(seg.getTrack());
            newGroup.add(dup);
        }
        return newGroup;
    }
}

File: core/src/main/scala/com/lomekwi/cave/timeline/SegmentGroup.scala
package com.lomekwi.cave.timeline

import com.lomekwi.cave.pipeline.Segment

import java.io.Serializable
import java.util

/**
 * 一组被同时操作（拖动/分割）的片段。组可以跨轨道，因此不归属于任何轨道，
 * 由 [[Timeline]] 统一登记。
 */
@SerialVersionUID(1L)
class SegmentGroup extends util.AbstractCollection[Segment[?]] with Serializable {
  private final val segments: util.Set[Segment[?]] = new util.LinkedHashSet[Segment[?]]()

  override def add(segment: Segment[?]): Boolean = {
    segments.add(segment)
  }

  override def remove(o: Any): Boolean = {
    segments.remove(o)
  }

  override def clear(): Unit = {
    segments.clear()
  }

  override def contains(o: Any): Boolean = {
    segments.contains(o)
  }

  override def isEmpty: Boolean = {
    segments.isEmpty
  }

  override def iterator(): util.Iterator[Segment[?]] = {
    segments.iterator()
  }

  override def size(): Int = {
    segments.size()
  }
}
